### PR TITLE
CCCP-3646: Update the code to consider categories update

### PR DESCRIPTION
**Pull request by miloudbagaa**

### CODE REVIEW

File: tasks/idrac_update.yml [hunk 1]
```diff
@@ -0,0 +1,27 @@
+---
+
+    - name: Check if dsu is installed
```
Comment (Klaas-) [inline tasks/idrac_update.yml]: You do not need to check for that, https://github.com/CSCfi/ansible-role-dell-firmware-upgrade/blob/e5aa03dcfb56a849e8275b02f39840a864824d8c/tasks/main.yml#L61-L64 is making sure dsu exists

File: tasks/idrac_update.yml [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+      shell: command -v dsu
+      register: output_dsu_commad
+      failed_when: ( output_dsu_commad.stderr | length  > 0 )
+
+    - name: Set is_dsu_installed variable
+      set_fact:
+        is_dsu_installed: "{{ output_dsu_commad.stdout | length  > 0  }}"
+
+    - name: Upgrade the iDRAC if applicable
```
Comment (Klaas-) [inline tasks/idrac_update.yml]: I also don't think you need to do that here, you can most likely add an option to dell_dsu_arguments that is conditional on upgrade_idrac_only

File: tasks/idrac_update.yml [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+    - name: Upgrade the iDRAC if applicable 
+      shell: dsu --category=DR --non-interactive --apply-upgrades
+      register: idrac_update_output
+      failed_when: ( idrac_update_output.stderr | length  > 0 )
+      when: (is_dsu_installed == True)
+
+    - name: Check if the upgrades are applied
+      set_fact:
+        is_dsu_upgraded: "{{ 'Installed successfully' in idrac_update_output.stdout }}"
```
Comment (Klaas-) [inline tasks/idrac_update.yml]: I don't think you need this task file at all -- but just for completeness: look at how it's done in the other file - we check the return codes not the output.

File: tasks/idrac_update.yml (hunk 2 above)
Comment (miloudbagaa) [reply tasks/idrac_update.yml]: I agree. I am updating the code to be generic and update any components instead only iDrac.

File: defaults/main.yml [hunk 4]
```diff
@@ -32,4 +32,24 @@ dell_dsu_changed_exit_codes:
   - 8 # Reboot is required for the update to be completed successfully
 dell_dsu_not_changed_exit_codes:
   - 34 # There were no updates found for the system.
+
+firmeware_components:
```
Comment (junousi) [inline defaults/main.yml]: While this doesn't apply to everything, I think this is a case where we need to be pedantic.

`firmeware` -> `firmware`

**Review by junousi — DISMISSED** [review]

File: defaults/main.yml (hunk 4 above)
Comment (miloudbagaa) [reply defaults/main.yml]: I agree. I will update it

File: tests/test.yml [hunk 5]
```diff
@@ -5,7 +5,7 @@
    vars:
      - dell_dsu_repo_install: True
      - dell_dsu_update_all_firmware: False
-     - dell_dsu_arguments: "--non-interactive --apply-upgrades-only --preview"
+     - dell_dsu_arguments: "--non-interactive --apply-upgrades --preview"
```
Comment (junousi) [inline tests/test.yml]: In a way this PR is changing the role's "user interface" from

`dell_dsu_arguments`

to the tuple of

`dell_dsu_arguments_base`
`upgrade_categories`

so we could almost say that `dell_dsu_arguments` becomes an internal variable of the role. Thus I think the tests should also be modified to utilize the tuple. (And e.g. default to something simple like `upgrade_categories="iDRAC"`)

Naturally the Travis Docker container in which the test runs doesn't have Dell hardware, but I suppose that shouldn't matter when running `--preview` and `dell_dsu_assert:False`.

File: defaults/main.yml [hunk 6]
```diff
@@ -21,7 +21,7 @@ dell_dsu_gpg_keys:
 # Arguments to DSU
 # Could be changed to for eaxmple only upgrade some categories with "--category=BI"
```
Comment (junousi) [inline defaults/main.yml]: Heyyyy I think this PR changes the "could be changed" to "done!" :) You can remove this comment.

File: tasks/main.yml [hunk 7]
```diff
@@ -63,9 +63,55 @@
     name: dell-system-update
     state: latest
 
+- name: Init dell_dsu_update_all_firmware
+  set_fact:
+    dell_dsu_update_all_firmware: False
```
Comment (junousi) [inline tasks/main.yml]: I think this breaks workflow for users who have standardized on running stuff with CLI args `-e dell_dsu_update_all_firmware=True`.

Trying to wrap my brain around this. In general I think this functionality could be implemented *without* overriding the variable with set_fact, and primarily do what the user is used to. Rather than introducing a lot of changes in how the role is used.

Comment (Klaas-) [reply tasks/main.yml]: can you just use inline magic when defaulting `dell_dsu_arguments: "--non-interactive --apply-upgrades"`?
you know like `dell_dsu_arguments: "--non-interactive --apply-upgrades {{ categories_list if categories_list is defined }}"`
with some more magic to make the list a comma seperated string with the right option in front of it? That way you can still run checks on the list if defined but keep the logic of the role mostly the same if it's not used

Comment (miloudbagaa) [reply tasks/main.yml]: It is not possible to use this dell_dsu_arguments: "--non-interactive --apply-upgrades {{ categories_list if categories_list is defined }}". The user will include names, such as BIOS,iDRAC...  Then, these names should be converted to symbols BI,DR... using the dictionary firmware_components.

Comment (Klaas-) [reply tasks/main.yml]: well you could also do a transformation in the default definition it will become fairly difficult to read inline magic then :) but you could also just have the users specify the proper component names for dsu without transforming them. I mean to me as a user there is not much of a difference of specifying iDRAC or DR directly.

Comment (Klaas-) [reply tasks/main.yml]: `dell_dsu_arguments: "--non-interactive --apply-upgrades{% if dsu_categories is defined %} --category={% set comma = joiner(',') %}{% for dsu_category in dsu_categories %}{{ comma() }}{{ firmware_components[dsu_category] }}{% endfor %}{% endif %}"`
something like this :)

Comment (miloudbagaa) [reply tasks/main.yml]: Thanks. I will check it by tomorrow and see.

Comment (miloudbagaa) [reply tasks/main.yml]: I have implemented your suggestions. Now, the code is shorter :).

Comment (Klaas-) [reply tasks/main.yml]: you can even put it into defaults/main.yml that would make it even easier and you'd need less checks. You only need to move the firmware_components definition in front of the dell_dsu_arguments definition

Comment (miloudbagaa) [reply tasks/main.yml]: Sorry, I did not get it. I need  categories_list to be defined before creating dell_dsu_arguments. In some cases, we need just to install the dsu without any updates.

Comment (Klaas-) [reply tasks/main.yml]: I think we're having a mismatch on what you want to achieve :) I thought you only want to add the possibility to upgrade certain types of firmware. Having this role not execute dsu upgrade, only installing it, was not on my scope

What I would do - create two input variables:
dsu_install_firmware (or something like that, default it to true)
dsu_categories, list of firmware components you want to upgrade, undefined by default. Can be defined by user as BIOS,iDRAC etc needs to be a list

Now you can use defaults/main.yml to do most of the work, something like what I already posted
```
dell_dsu_arguments: "--non-interactive --apply-upgrades{% if dsu_categories is defined %} --category={% set comma = joiner(',') %}{% for dsu_category in dsu_categories %}{{ comma() }}{{ firmware_components[dsu_category] }}{% endfor %}{% endif %}"
dsu_install_firmware: True
```

And in tasks/main.yml you only need to change the when clause of the update to 
`when: dsu_install_firmware | bool`

* this will keep the default behavior of update all firmware when dsu_categories is not supplied (ie no --category on dsu cli)
* it will change behavior when dsu_categories  is supplied to pass  --category=BI,... to dsu cli
* if you overwrite dsu_install_firmware: False it will not run dsu

I did not syntax check it, but at least I think it should work and minimize the changes you need to make :)

Comment (miloudbagaa) [reply tasks/main.yml]: Great. I will see how I can implement this changes.

Comment (miloudbagaa) [reply tasks/main.yml]: We are trying to update the code while keeping original code. As the code is already used by many users. 

For more context:

1) The code should install only dsu if no parameter has been defined. The ansible playbook that use this role, it sets dell_dsu_update_all_firmware to False. 

2) The code should update all the firmware if one of the following holds: i) dell_dsu_update_all_firmware=True; ii) upgrade_categories="all" 

3) The code should update only some firmware components if upgrade_categories is not set to all and has been defined with some components.  We have to assert to ensure that the passed parameters are valid. For example if we plan to update only BIOS and iDRAC, we just specify upgrade_categories="iDRAC,BIOS".

I think that the code that we have now is optimized for these use-cases and keep the old version working for others.

Comment (Klaas-) [reply tasks/main.yml]: I think this makes the role more complicated :) Anyways, I think the current code should work, but I think you need to check if you are making a breaking change, I think you can no longer overwrite dell_dsu_arguments which you could before your change (because you use set_fact to set dell_dsu_arguments)

File: tasks/main.yml [hunk 8]
```diff
@@ -63,9 +63,38 @@
...
+  when:
+    - upgrade_categories is not defined
+    - (dell_dsu_update_all_firmware is defined) and (dell_dsu_update_all_firmware)
+
+- name: Set categories_list
+  set_fact:
+    categories_list: "{{ upgrade_categories.split(',') }}"
+  when: (upgrade_categories is defined)
+
+- name: Set dell_dsu_update_all_firmware to True
```
Comment (Klaas-) [inline tasks/main.yml]: Should this be defined before you use it in line 71?

Comment (miloudbagaa) [reply tasks/main.yml]: No. It was not there before my update. This is the original code: https://github.com/CSCfi/ansible-role-dell-firmware-upgrade/blob/master/tasks/main.yml

No need to init dell_dsu_update_all_firmware to True as by default has the value true. Please check this:
https://github.com/CSCfi/ansible-role-dell-firmware-upgrade/blob/master/defaults/main.yml

File: tasks/main.yml (hunk 7 above)
Comment (miloudbagaa) [reply tasks/main.yml]: Before my change. dell_dsu_arguments has a fixed only one value, which is "--non-interactive --apply-upgrades". Now, it has more granularity. 

In the current code, we have dell_dsu_update_all_firmware = True. So, thanks to tasks defined from line 66 to 92, dell_dsu_arguments will take the value "--non-interactive --apply-upgrades" by default. Otherwise, if upgrade_categories is defined with more granularity, such as upgrade_categories="BIOS,iDRAC", then  dell_dsu_arguments will be fixed to "--non-interactive --apply-upgrades --category=BI,DR".

File: tasks/main.yml (hunk 8 above)
Comment (Klaas-) [reply tasks/main.yml]: I did not see that. But then the changes you make even make less sense to me :)

if it defaults to true, why set it to true again `when: (categories_list is defined) and ("all" in categories_list) and (categories_list | length == 1)` you would just not set catagories list if you want to update all firmware? :)

File: tasks/main.yml (hunk 7 above)
Comment (Klaas-) [reply tasks/main.yml]: > Before my change. dell_dsu_arguments has a fixed only one value, which is "--non-interactive --apply-upgrades". Now, it has more granularity.

no, I think you're missing some ansible basics. you can overwrite anything in defaults/main.yml without any code in tasks/main.yml, so you could overwrite it before with _anything_ you want including additional categories etc. For example I used it to add auto accept of gpg keys before the role implemented that. 

tldr: everything in defaults/main.yml is not a fixed value, it is a default that you can overwrite when calling the role or via host vars or ....

File: tasks/main.yml (hunk 8 above)
Comment (miloudbagaa) [reply tasks/main.yml]: I agree. The value of dell_dsu_update_all_firmware is True, so there is no need to set it again true. However, in our ansible playbook, we set dell_dsu_update_all_firmware to False. The idea to set dell_dsu_update_all_firmware when categories_list is set to "all". However, to remove the confusion, I plan to remove this option.

The farmware can be partially update if categories_list is defined with a list. Otherwise, all the framware should be updated if dell_dsu_update_all_firmware is True.

File: tasks/main.yml (hunk 7 above)
Comment (miloudbagaa) [reply tasks/main.yml]: Yes. I know that. The default value can be overwritten within the task or by passed parameters.

File: tasks/main.yml (hunk 8 above)
Comment (Klaas-) [reply tasks/main.yml]: well the default case is nothing is overwritten, just defaults defined, in that case it should update by running "dsu --non-interactive --apply-upgrades" I took another look at your code and I don#t think it does that currently. But I only read the code, so I may have missed something

--> https://github.com/CSCfi/ansible-role-dell-firmware-upgrade/blob/386ce3198d04df4902dd88eca0b52c1e3d2c939c/tasks/main.yml#L66-L71
not executed because upgrade_categories is undefined
--> https://github.com/CSCfi/ansible-role-dell-firmware-upgrade/blob/386ce3198d04df4902dd88eca0b52c1e3d2c939c/tasks/main.yml#L73-L76
not executed  because upgrade_categories is undefined
--> https://github.com/CSCfi/ansible-role-dell-firmware-upgrade/blob/386ce3198d04df4902dd88eca0b52c1e3d2c939c/tasks/main.yml#L78-L81
not executed because categories_list is undefined
--> https://github.com/CSCfi/ansible-role-dell-firmware-upgrade/blob/386ce3198d04df4902dd88eca0b52c1e3d2c939c/tasks/main.yml#L83-L87
not executed because categories_list is undefined
--> https://github.com/CSCfi/ansible-role-dell-firmware-upgrade/blob/386ce3198d04df4902dd88eca0b52c1e3d2c939c/tasks/main.yml#L89-L92
not executed because categories_list  is undefined
--> https://github.com/CSCfi/ansible-role-dell-firmware-upgrade/blob/386ce3198d04df4902dd88eca0b52c1e3d2c939c/tasks/main.yml#L94-L99
not executed because dell_dsu_arguments is not defined

Comment (miloudbagaa) [reply tasks/main.yml]: I have simplified the code by don't considering upgrade_categories="all". Now, the code does the following:

1) Update all the firmware if dell_dsu_update_all_firmware=True
2) Otherwise, it updates only some components if upgrade_categories is defined. For instance if upgrade_categories="BIOS,iDRAC", then only these two components should be updated.

Comment (Klaas-) [reply tasks/main.yml]: I think this should work now :) I did not test it, only read the code, so maybe give it a couple of spins with the different scenarios on a real dell server :)

Apart from that: I think you still need to update the readme with the new feature (being able to set upgrade_categories to a string of comma seperated categories that need to match the hash keys in firmware_components) and you should make a note about the breaking change (not being able to overwrite dell_dsu_arguments anymore)

before merging, maybe squash the commits :)

Comment (miloudbagaa) [reply tasks/main.yml]: I already did the tests with our servers. 

I will update readme and push the changes.

Comment (miloudbagaa) [reply tasks/main.yml]: I have updated the readme by showing the new features.

File: README.md [hunk 9]
```diff
@@ -10,9 +10,11 @@ Role Variables
 
 See defaults/main.yml for details.
 
-By default this role does not install the Dell DSU yum repo.
+By default this role does not install the Dell DSU yum repo. To install Dell DSU, the variable dell_dsu_repo_install should be set to True.
 
-By default this role upgrades all firmwares.
+By default, this role upgrades all firmwares. To disable this functionality, the variable dell_dsu_update_all_firmware should be set to True.
```
Comment (Klaas-) [inline README.md]: I think you mean "set to False"

Comment (miloudbagaa) [reply README.md]: Yes. A typos.

File: README.md [hunk 10]
```diff
@@ -30,12 +32,31 @@ Dependencies
...
+* If you plan to do not upgrade all firmwares, then you can use the following code: 
 ```
 - hosts: servers
+  vars:
+    - dell_dsu_update_all_firmware: False
   roles:
      - { role: ansible-role-dell-firmware-upgrade }
 ```
+
+* If you plan to upgrade only iDRAC and BIOS firmwares, then you can use the following code:
```
Comment (Klaas-) [inline README.md:51]: I would add a link to defaults/main.yml#L34-L52 for the possible components (or just a note about looking them up in that file)

Comment (miloudbagaa) [reply README.md:51]: I agree. Thanks

**Review by Klaas- — APPROVED** [review]
I think this is good now :)

**Comment by miloudbagaa** [comment]
Thanks Klaas for your deep dive review. The code looks better now.

**Comment by Klaas-** [comment]
> Thanks Klaas for your deep dive review. The code looks better now.

no problem, I'll give it a spin on my dell servers in the next couple of weeks

File: README.md [hunk 11]
```diff
@@ -10,9 +10,11 @@ Role Variables
 
 See defaults/main.yml for details.
 
-By default this role does not install the Dell DSU yum repo.
+By default this role does not install the Dell DSU yum repo. To install Dell DSU, the variable dell_dsu_repo_install should be set to True.
 
-By default this role upgrades all firmwares.
+By default, this role upgrades all firmwares. To disable this functionality, the variable dell_dsu_update_all_firmware should be set to False.
+
+If some firmwares are defined in upgrade_categories, then only these firmwares should be upgrated. If upgrade_categories is defined, then dell_dsu_update_all_firmware is ignored.
```
Comment (Aegil) [inline README.md]: upgrated/upgraded

File: README.md [hunk 12]
```diff
@@ -30,12 +32,33 @@ Dependencies
 Example Playbook
 ----------------
 
-* You can simply use this role like
+* You can simply use this role like. In this case, the code will upgrade all firmwares.
```
Comment (Aegil) [inline README.md]: role like below.

File: README.md (hunk 11 above)
Comment (bagmoul) [reply README.md]: Thanks

**Review by Aegil — APPROVED** [review]
